dkist/wcs/models.py

```diff
@@ -0,0 +1,329 @@
+    translation = (0, 0)
+    if spatial_unit is not None:
+        translation *= pc.unit
+        # If we have units then we need to convert all things to Quantity
+        # as they might be Parameter classes
+        crpix = u.Quantity(crpix)
+        cdelt = u.Quantity(cdelt)
+        crval = u.Quantity(crval)
+        lon_pole = u.Quantity(lon_pole)
+        pc = u.Quantity(pc)
+
+    shift = m.Shift(-crpix[0]) & m.Shift(-crpix[1])
+    scale = m.Multiply(cdelt[0]) & m.Multiply(cdelt[1])
+    rot = m.AffineTransformation2D(pc, translation=translation)
+    skyrot = m.RotateNative2Celestial(crval[0], crval[1], lon_pole)
+    return shift | scale | rot | projection | skyrot
+
+
+class BaseVaryingCelestialTransform(Model):
+    """
+    Shared components between the forward and reverse varying celestial transforms.
+    """
+
+    standard_broadcasting = False
+    _separable = False
+    _input_units_allow_dimensionless = True
+
+    crpix = Parameter()
+    cdelt = Parameter()
+    lon_pole = Parameter(default=180)
+
+    @staticmethod
+    def _validate_table_shapes(pc_table, crval_table):
+        table_shape = None
+        if pc_table.shape != (2, 2):
+            if pc_table.shape[-2:] != (2, 2):
+                raise ValueError("The pc table should be an array of 2x2 matrices.")
+            table_shape = pc_table.shape[:-2]
+
+        if crval_table.shape != (2,):
+            if crval_table.shape[-1] != 2:
+                raise ValueError("The crval table should be an array of coordinate "
+                                 "pairs (the last dimension should have length 2).")
+
+            if table_shape is not None:
+                if table_shape != crval_table.shape[:-1]:
+                    raise ValueError("The shape of the pc and crval tables should match. "
+                                     f"The pc table has shape {table_shape} and the "
+                                     f"crval table has shape {crval_table.shape[:-1]}")
+            table_shape = crval_table.shape[:-1]
+
+        return table_shape
+
+    @staticmethod
+    def get_pc_crval(z, pc, crval):
+        # Get crval and pc
+        if isinstance(z, u.Quantity):
+            ind = int(z.value)
+        else:
+            ind = int(z)
+        if pc.shape != (2, 2):
+            pc = pc[ind]
+        if crval.shape != (2,):
+            crval = crval[ind]
+        return pc, crval
+
+    def __init__(self, *args, crval_table=None, pc_table=None, projection=m.Pix2Sky_TAN(), **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.pc_table = np.asanyarray(pc_table)
+        self.crval_table = np.asanyarray(crval_table)
+        self.table_shape = self._validate_table_shapes(self.pc_table, self.crval_table)
+
+        if not isinstance(projection, m.Pix2SkyProjection):
+            raise TypeError("The projection keyword should be a Pix2SkyProjection model class.")
+        self.projection = projection
```

Comment:
Just a caution comment here - if any projection is allowed then care should be taken to compute `lon_pole`. This is not trivial in the general case. There should be a function in gwcs "soon" which retrieves the value from wcslib.

Reply:
I don't think we will be using it for any projections other than TAN, but worth keeping an eye on thanks.